src/backend/distributed/planner/intermediate_result_pruning.c

```diff
@@ -0,0 +1,257 @@
+															 distributedPlan,
+															 workerNodeCount);
+
+			elog(DEBUG4, "subplan %s is used in %lu", resultId, distributedPlan->planId);
+		}
+	}
+
+	/* descend into the subPlans */
+	foreach(subPlanCell, subPlanList)
+	{
+		DistributedSubPlan *subPlan = (DistributedSubPlan *) lfirst(subPlanCell);
+		CustomScan *customScan = FetchCitusCustomScanIfExists(subPlan->plan->planTree);
+		if (customScan)
+		{
+			DistributedPlan *distributedPlanOfSubPlan = GetDistributedPlan(customScan);
+			RecordSubplanExecutionsOnNodes(intermediateResultsHash,
+										   distributedPlanOfSubPlan);
+		}
+	}
+}
+
+
+/*
+ * AppendAllAccessedWorkerNodes iterates over all the tasks in a distributed plan
+ * to create the list of worker nodes that can be accessed when this plan is executed.
+ *
+ * If there are multiple placements of a Shard, all of them are considered.
+ */
+static List *
+AppendAllAccessedWorkerNodes(List *workerNodeList, DistributedPlan *distributedPlan, int
+							 workerNodeCount)
+{
+	List *taskList = distributedPlan->workerJob->taskList;
+	ListCell *taskCell = NULL;
+
+	foreach(taskCell, taskList)
+	{
+		Task *task = lfirst(taskCell);
+		ListCell *placementCell = NULL;
+		foreach(placementCell, task->taskPlacementList)
+		{
+			ShardPlacement *placement = lfirst(placementCell);
+			workerNodeList = list_append_unique_int(workerNodeList, placement->nodeId);
+
+			/* early return if all the workers are accessed */
+			if (list_length(workerNodeList) == workerNodeCount)
+			{
+				return workerNodeList;
+			}
+		}
+	}
+
+	return workerNodeList;
+}
+
+
+/*
+ * MakeIntermediateResultHTAB is a helper method that creates a Hash Table that store information on the intermediate result.
+ */
+HTAB *
+MakeIntermediateResultHTAB()
+{
+	HTAB *intermediateResultsHash = NULL;
+	uint32 hashFlags = 0;
+	HASHCTL info = { 0 };
+	int initialNumberOfElements = 16;
+
+	info.keysize = sizeof(IntermediateResultsHashKey);
+	info.entrysize = sizeof(IntermediateResultsHashEntry);
+	info.hash = string_hash;
+	info.hcxt = CurrentMemoryContext;
+	hashFlags = (HASH_ELEM | HASH_FUNCTION | HASH_CONTEXT);
+
+	intermediateResultsHash = hash_create("Intermediate results hash",
+										  initialNumberOfElements, &info, hashFlags);
+
+	return intermediateResultsHash;
+}
+
+
+/*
+ * FindAllWorkerNodesUsingSubplan creates a list of worker nodes that
+ * may need to access subplan results.
+ */
+List *
+FindAllWorkerNodesUsingSubplan(HTAB *intermediateResultsHash,
+							   char *resultId)
+{
+	List *workerNodeList = NIL;
+	IntermediateResultsHashEntry *entry =
+		SearchIntermediateResult(intermediateResultsHash, resultId);
+
+	ListCell *nodeIdCell = NULL;
+	foreach(nodeIdCell, entry->nodeIdList)
+	{
+		WorkerNode *workerNode = LookupNodeByNodeId(lfirst_int(nodeIdCell));
+
+		workerNodeList = lappend(workerNodeList, workerNode);
+		elog(DEBUG1, "%s is broadcasted to %s:%d", resultId, workerNode->workerName,
```

Comment:
Maybe I am missing a point but at this point, are the results broadcasted?

Reply:
This is actually not the part of code that does the broadcasting. Here we only determine which worker will receive the intermediate results as this is part of distributed planner.

It might have been more appropriate to say `will be broadcasted to` or `will be multicasted to` here